compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/calc/MulNode.java

```diff
@@ -92,27 +93,37 @@ public ValueNode canonical(CanonicalizerTool tool, ValueNode forX, ValueNode for
         }
         BinaryOp<Mul> op = getOp(forX, forY);
         NodeView view = NodeView.from(tool);
-        return canonical(this, op, stamp(view), forX, forY, view);
+        return canonical(tool, this, op, stamp(view), forX, forY, view);
     }
 
-    private static ValueNode canonical(MulNode self, BinaryOp<Mul> op, Stamp stamp, ValueNode forX, ValueNode forY, NodeView view) {
+    private static ValueNode canonical(CanonicalizerTool tool, MulNode self, BinaryOp<Mul> op, Stamp stamp, ValueNode forX, ValueNode forY, NodeView view) {
         if (forY.isConstant()) {
             Constant c = forY.asConstant();
             if (op.isNeutral(c)) {
                 return forX;
             }
 
-            if (c instanceof PrimitiveConstant && ((PrimitiveConstant) c).getJavaKind().isNumericInteger()) {
-                long i = ((PrimitiveConstant) c).asLong();
-                ValueNode result = canonical(stamp, forX, i, view);
-                if (result != null) {
-                    return result;
+            if (op.isAssociative()) {
+                // Canonicalize expressions like "(a * 2) * 4" => "(a * 8)"
+                ValueNode reassociated = reassociate1(self != null ? self : (MulNode) new MulNode(forX, forY).maybeCommuteInputs(), ValueNode.isConstantPredicate(), forX, forY, view);
```

Comment:
We try to always use factory methods of nodes so any immediate folding can occur during construction already

Reply:
Yes, it's better to use the `create` function normally. However, we cannot use the factory method here, since it is inside the `canonical` function which will also be called by "create". Using `create` here will create a dead loop, which looks like `create` -> `canonical` -> `create` -> `canonical` -> ...

Reply:
Yes, it's better to use the `create` normally. However, it cannot be used here, since this is inside of the `canonical` which is called by `create`. Using `create` here will make it a dead loop which looks like `create` -> `canonical` -> `create` -> `canonical` -> ...

And this is the log detail:
```
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.calc.MulNode.create(MulNode.java:70)
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.calc.MulNode.canonical(MulNode.java:108)
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.calc.MulNode.create(MulNode.java:70)
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.calc.MulNode.canonical(MulNode.java:108)
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.calc.MulNode.create(MulNode.java:70)
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.calc.MulNode.canonical(MulNode.java:108)
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.calc.MulNode.create(MulNode.java:70)
```

Reply:
ah, yes overlooked that ;-) never mind